1.21.x wind rework

alekiponi opened the pull request:
- change wind to use a new scale and a new distribution
 - the new range is from 0 to 32 m/s or 115 km/s
 - 0 to 8 m/s uses a normal distribution, 8 to 32 m/s are more rare and only happen at a certain threshold
 - rain pushes the mean up by 3.2 m/s
 - altitude adds to wind speed
- makes windmills respect wind speed
- adjustments to wind-affected particles
- make windmills and water wheels stop instead of breaking when obstructed
- add a config to change how rotational speed is displayed
- make the default unit for rotational speed RPM, a much more standard unit
- prevent windmills from rotating with only one blade
- add JADE/TOP tooltips to inform players of problems with windmills and water wheels

Commit:
- change wind to use a new scale and a new distribution
- make windmills respect wind speed
- adjustments to wind-affected particles

Commit:
revert random helpers change

alekiponi committed:
Merge branch 'TerraFirmaCraft:1.21.x' into 1.21.x-wind-rework

Commit:
minor change to how wind affects windmills

Commit:
Merge remote-tracking branch 'origin/1.21.x-wind-rework' into 1.21.x-wind-rework

eerussianguy commented on src/main/java/net/dries007/tfc/client/ClientForgeEventHandler.java:
```diff
@@ -461,7 +461,7 @@ public static void onScreenKey(ScreenEvent.KeyPressed.Pre event)
             if (slot != null)
             {
-               PacketDistributor.sendToServer(new StackFoodPacket(slot.index));
+                PacketDistributor.sendToServer(new StackFoodPacket(slot.index));
```
why are there random whitespace cahnges everywhere

eerussianguy commented on src/main/java/net/dries007/tfc/common/blockentities/rotation/WindmillBlockEntity.java:
```diff
@@ -38,20 +41,34 @@ public class WindmillBlockEntity extends TickableInventoryBlockEntity<ItemStackH
+            if (isObstructedBySolidBlocks(level, pos, state.getValue(WindmillBlock.AXIS))){
+                // Check every two seconds if the windmill is obstructed, and if so, break
+                level.destroyBlock(pos, true);
+            }
```
braces on newlines

eerussianguy replied:
Also I'm confused why this is being added. We handle the obstruction check in `clientTick`. By that being called from `serverTick`, we are handling that logic on the server at the correct time. If that doesn't work, find out why. Plus, doing all this logic and *then* checking if the windmill should be destroyed is an antipattern.

eerussianguy commented on src/main/java/net/dries007/tfc/common/blockentities/rotation/WindmillBlockEntity.java:
```diff
@@ -38,20 +41,34 @@ public class WindmillBlockEntity extends TickableInventoryBlockEntity<ItemStackH
+
+            float windFactor = Math.min(wind, 0.5f) + 0.75f; // clamp below ~57 kmh
+
+            windmill.targetSpeed = windFactor * targetBeforeWind;
```
So here's the logic flow:

- Client is spinning the windmill at speed X
- Up to 2s later, Server calculates target speed Y
- Server sends a sync packet regardless of if Y=X or not
- Client is now spinning at Y
- Client's 'displayed speed' quickly lerps to meet Y

Two problems:
- I don't see why the server needs to calc-then-sync. Wind is available on the client and we can simply run the same calc on either side to keep things up to date. The server needs to know about rotational speed for the purpose of mechanical power. The client needs it for mech. and rendering (which is why it has the lerping functionality that the server doesn't). This is sort of a mindset I've tried to get out of, where we put the server in charge of things that should just be ran on both sides. The reason is that it causes unnecessary syncing, and also is a desync because the server experiences things before the client does.
- This is sending a packet every 2s regardless of what's happening, and we did not need to do that before. Sending constant packets can be a major lag source. I'm not totally sure why this was necessary to add. If it is needed then there should at least be checks that the syncing is necessary.

eerussianguy commented on src/main/java/net/dries007/tfc/util/tracker/WeatherHelpers.java:
```diff
@@ -600,4 +603,13 @@ public static boolean isIce(BlockState state)
+
+    public static float windKMS(Vec2 wind){
+        return wind.length() * WIND_KMS_FACTOR;
+    }
```
braces on newlinesplease

Commit:
- make windmills do all calculations on both sides and only sync when necessary
- make windmills and water wheels stop instead of breaking when obstructed
- add a config to change how rotational speed is displayed
- make the default unit for rotational speed RPM, a much more standard unit
- prevent windmills from rotating with only one blade
- add JADE/TOP tooltips to inform players of problems with windmills and water wheels

Commit:
change flow to do less unnecessary stuff in windmill and water wheel tick

Commit:
update resources

Commit:
add a helper to convert to M/tick

eerussianguy commented on src/main/java/net/dries007/tfc/common/blockentities/rotation/WindmillBlockEntity.java:
```diff
@@ -34,40 +37,66 @@ public class WindmillBlockEntity extends TickableInventoryBlockEntity<ItemStackH
         {
-            // Check every two seconds if the windmill is obstructed, and if so, break
-            level.destroyBlock(pos, true);
+            if (!windmill.obstructed && isObstructedBySolidBlocks(level, pos, state.getValue(WindmillBlock.AXIS)))
+            {
+                windmill.obstructed = true;
+                windmill.markForSync();
+            }
+            if (windmill.obstructed && !isObstructedBySolidBlocks(level, pos, state.getValue(WindmillBlock.AXIS)))
+            {
+                windmill.obstructed = false;
+                windmill.markForSync();
+            }
```
```java
boolean obstructedNow = isObstructedBySolidBlocks();
if (obstructedNow != windmill.obstructed)
{
    windmill.obstructed = obstructedNow;
    windmill.markForSync();
}
```

eerussianguy commented on src/main/java/net/dries007/tfc/config/ClientConfig.java:
```diff
@@ -117,6 +118,14 @@ public class ClientConfig extends BaseConfig
             "  LEFT_HOTBAR = Move elements closer to the hotbar; when fishing or riding a jumping entity, those elements will appear as a vertical bar between the hotbar and offhand slot"
         ).define("disabledExperienceBarStyle", DisabledExperienceBarStyle.HOVER);
 
+        rotationDisplayStyle = builder.comment(
+            "Changes the units that rotational speed is displayed in.",
+            "  ROTATIONS_PER_SECOND - Shows rotational speed as a measure of full rotations per second",
+            "  DEGREES_PER_SECOND - Shows rotational speed as a measure of degrees per second",
+            "  REVOLUTIONS_PER_SECOND - Shows rotational speed as a measure of revolutions per second",
+            "  REVOLUTIONS_PER_MINUTE - Shows rotational speed as a measure of revolutions per minute"
+        ).define("rotationDisplayStyle", RotationDisplayStyle.REVOLUTIONS_PER_MINUTE);
+
```
I think these might be lowercase-ized in the actual config? check please

alekiponi replied:
This is correct as far as I can tell

eerussianguy commented on src/main/java/net/dries007/tfc/util/tooltip/BlockEntityTooltips.java:
```diff
@@ -205,10 +227,7 @@ public static void register(RegisterCallback<BlockEntityTooltip, Block> callback
     public static final BlockEntityTooltip BELLOWS = (level, state, pos, entity, tooltip) -> {
         if (entity instanceof BellowsBlockEntity bellows)
         {
-            final int step = Mth.clamp((int) Mth.map(
-                bellows.getExtensionLength(0f),
-                BellowsBlockEntity.MIN_EXTENSION, BellowsBlockEntity.MAX_EXTENSION,
-                0, 10), 0, 10);
+            final int step = Mth.clamp((int) Mth.map(bellows.getExtensionLength(0f), BellowsBlockEntity.MIN_EXTENSION, BellowsBlockEntity.MAX_EXTENSION, 0, 10), 0, 10);
```
why was this changed?

alekiponi replied:
auto format

eerussianguy commented on src/main/java/net/dries007/tfc/util/tooltip/BlockEntityTooltips.java:
```diff
@@ -504,12 +523,43 @@ else if (pot.getOutput() != null && !pot.getOutput().isEmpty())
         {
             for (ItemStack stack : Helpers.iterate(placedItem.getInventory()))
             {
-                if (!stack.isEmpty())
-                    tooltip.accept(stack.getHoverName());
+                if (!stack.isEmpty()) tooltip.accept(stack.getHoverName());
```
why was this changed?

alekiponi replied:
auto format

eerussianguy commented on src/main/java/net/dries007/tfc/util/tracker/WeatherHelpers.java:
```diff
@@ -600,4 +603,26 @@ public static boolean isIce(BlockState state)
+    /**
+     * Converts wind speed to M/s
+     */
+    public static float windMS(Vec2 wind){
+        return wind.length() * WIND_MS_FACTOR;
+    }
+
+    /**
+     * Converts wind speed to M/tick (useful for accurately affecting entity/particle velocity)
+     */
+    public static float windMT(Vec2 wind){
+        return wind.length() * WIND_MS_FACTOR / 20;
+    }
+
+    /**
+     * Converts wind speed to KM/h
+     */
+    public static float windKMS(Vec2 wind){
+        return wind.length() * WIND_KMS_FACTOR;
+    }
```
braces on newlines

alekiponi replied:
forgot to auto format

Commit:
minor fixes, add translation for rotation config

Commit:
fix typo in lang

Commit:
resources

Commit:
spawn wind particles at slower wind speeds
